### ISSUE TITLE
`App::configure_sets` should work the same way as `.before` and `.after`

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#13064]
## What problem does this solve or what need does it fill?

[`configure_sets`](https://docs.rs/bevy/latest/bevy/app/struct.App.html#method.configure_sets) takes an `impl [IntoSystemSetConfigs](https://docs.rs/bevy/latest/bevy/ecs/schedule/trait.IntoSystemSetConfigs.html)` as its argument.

[`before`](https://docs.rs/bevy/latest/bevy/ecs/prelude/trait.IntoSystemConfigs.html#method.before) and similar methods take an [`impl IntoSystemSet`](https://docs.rs/bevy/latest/bevy/ecs/schedule/trait.IntoSystemSet.html).

Functionally, these are identical, except that system-type labels (e.g. `my_system`) don't work with the former.
This distinction is deeply unclear to users, who will simply get confused and frustrated about why the "obvious" thing doesn't work.

## What solution would you like?

Swap to using `IntoSystemSet` everywhere and unify the APIs.

This is both more consistent for users, and simpler to maintain, as we no longer need to maintain this distinction.

## What alternative(s) have you considered?

I encountered this in the context of attempting to order `apply_state_transition` systems relative to each other. Because they don't expose a public label, I can't do this directly.

However I *could* create a dummy system which does nothing, and order it relative to the two systems I actually care about, thus inducing a transitive ordering.

This is a dumb workaround: we're not actually protecting users in any meaningful way.

## Additional context

If I recall correctly, @maniwani was originally opposed to being allowed to configure system-type sets due to... concerns around being deliberate about public surface area for configuration?

Ultimately while I do think that's a reasonable concern, I don't think it's worth the inconsistencies and frustration for users here.

**Comment by hymm** [comment]
> If I recall correctly, @maniwani was originally opposed to being allowed to configure system-type sets due to... concerns around being deliberate about public surface area for configuration?

Yeah. Like if you do `my_system.after(apply_buffers)` that's almost certainly not what you actually want. I think this is less of a concern now that `apply_buffers` is less necessary. The other internal system that I can think of that gets added to a schedule multiple times is transform propagation. In that case users will probably adding their own set to order against since there are 2 systems responsible for propagating transforms.

**Comment by maniwani** [comment]
> If I recall correctly, @maniwani was originally opposed to being allowed to configure system-type sets due to... concerns around being deliberate about public surface area for configuration?

All instances of a given function will end up in the same system set, so if you have more than one instance of a system in the same schedule, configuring that set (or ordering on it) is ambiguous in a very subtle way. If you add system `foo` and system `bar` where `bar.after(foo)`, that's fine. However, if you add another `foo` to that same schedule the `bar.after(foo)` constraint is implicitly applied to that one too.

IMO now that `apply_deferred` (the most at-risk system) is automated, if we discourage having many instances of the same system in the same schedule, then removing the limits on system type set would be fine by me.

I think the problem would go away once all systems can be stored as entities, since there'd be no need to even have more than one instance of any system. Different schedules could grab the same instance (especially if we change the `Box` to an `Arc`).

**Comment by maniwani** [comment]
Also just want to mention that the schedule builder currently throws an error if you apply an ordering on system type sets that have multiple members.

**Comment by alice-i-cecile** [comment]
> Also just want to mention that the schedule builder currently throws an error if you apply an ordering on system type sets that have multiple members (because of the ambiguity).

Yep, I do like this safeguard, and think it should stay (at least for now). I think that with `apply_deferred` effectively no longer existing as a user-level concept + that safeguard we should be safe to relax the distinction and make `configure_sets` more consistent.
